mesonbuild/backend/backends.py

```diff
@@ -1484,3 +1484,17 @@ def get_introspection_data(self, target_id: str, target: build.Target) -> T.List
             }]
 
         return []
+
+    def get_devenv(self) -> build.EnvironmentVariables:
+        env = build.EnvironmentVariables()
+        if mesonlib.is_windows() or mesonlib.is_cygwin():
+            # On windows we cannot rely on rpath to run executables from build
+            # directory. We have to add in PATH the location of every DLL needed.
+            extra_paths = set()
+            for t in self.build.get_targets().values():
+                cross_built = not self.environment.machines.matches_build_machine(t.for_machine)
+                can_run = not cross_built or not self.environment.need_exe_wrapper()
+                if isinstance(t, build.Executable) and can_run:
+                    extra_paths.update(self.determine_windows_extra_paths(t, []))
```

Comment:
Please forgive my ignorance about Meson implementation details, but what exactly counts as `build.Executable` in Meson terms?

I wonder if there should be a way to opt out of this automatism, e.g. I don't know if we would want every little test/utility/example added to `PATH`.

Does it include all `library()`? If so, it would again be good to have an opt-out mechanism for plugins (which I don't think need to be in `PATH`?).

Reason this matters is that on Windows there's a hard limit on the number of characters for e.g. `PATH` of 4096 or somesuch and it's very easy to break that in GStreamer, for example, if your build directory is not underneath `C:\` but say `C:\Users\Jane Doe\Sources` (the current script is far from ideal in that respect).

Reply:
What that code does is look for every executable(), for each of them check every DLL it links from build directory and add those directories into PATH. It could be too much indeed... Probably should limit to installed exe.

Note that most of this logic is actually copied from gst-uninstalled, I'll double check what it does there for Windows path.

Reply:
@tp-m I checked gst-env.py, and as far as I can tell, it does the same thing. I guess if we hit Windows limitation there is not much we can do...

Now modified that code to also add in PATH the location of the executable itself, so they can be run from command line without specifying the path in build dir. gst-env.py does that for GStreamer tools.

Reply:
@tp-m FYI, I gave it a try in gst-build: https://gitlab.freedesktop.org/xclaesse/gst-build/-/commit/b829abdb1230dd96fb73b62989c8254bb01d4b76. With that I can run gst-inspect-1.0 and it list all plugins from my build directory, just like gst-env.py does.

There is still tones of hacks in gst-env.py that I did not understand yet, I'm sure we can copy what makes sense to improve Meson later.

Reply:
Yeah, just saying that there are problems with gst-env.py and things we could probably do better there.

It's cool to move this into Meson imho, makes a lot of sense to me.

Reply:
Do you have suggestions to make it better? Are there alternatives to setting huge PATH on Windows?

Reply:
My suggestion was to add a kwarg to `executable()` (or `subdir()`? but `executable()` makes more sense to me) to opt out of being added to the path automagically for the devenv. If this becomes part of Meson we can actually do this.

Reply:
`executable(..., devenv: false)` that default to the value of `install` kwarg? That would make sense indeed.

Another idea I had is `meson devenv` could first create `<builddir>/meson-sysroot` directory, copy all installed dll/exe there, and add just that directory into PATH. If we only copy when mtime changed (like `meson install` does) it should be fast enough. That would solve the problem completely, what do you think?

Reply:
Could do both actually, as @tristan957 mentioned below we could want to include some not installed executables into PATH too. In that case `executable(..., devenv: true)` could be handy.

Reply:
Now that I have thought about this more. I almost think that anything magical happening in the PR is just going to be difficult to maintain and that something like this could be much better:

```
meson.devenv(devenv, exectuables: [...])
```

Reply:
@tp-m let's start with including everything by default in the initial PR, we can extend later by adding a way to opt-out some exe, or workaround by copying into a sysroot dir, or any other fancy solution. OK?

Reply:
Sure :+1: